// @cloudflare/sandbox/dist/errors-BCXUmJUn.js
//#region ../shared/dist/errors/codes.js
/**
* Centralized error code registry
* Each code maps to a specific error type with consistent semantics
*/
const ErrorCode = {
	FILE_NOT_FOUND: "FILE_NOT_FOUND",
	PERMISSION_DENIED: "PERMISSION_DENIED",
	FILE_EXISTS: "FILE_EXISTS",
	IS_DIRECTORY: "IS_DIRECTORY",
	NOT_DIRECTORY: "NOT_DIRECTORY",
	NO_SPACE: "NO_SPACE",
	TOO_MANY_FILES: "TOO_MANY_FILES",
	RESOURCE_BUSY: "RESOURCE_BUSY",
	READ_ONLY: "READ_ONLY",
	NAME_TOO_LONG: "NAME_TOO_LONG",
	TOO_MANY_LINKS: "TOO_MANY_LINKS",
	FILESYSTEM_ERROR: "FILESYSTEM_ERROR",
	COMMAND_NOT_FOUND: "COMMAND_NOT_FOUND",
	COMMAND_PERMISSION_DENIED: "COMMAND_PERMISSION_DENIED",
	INVALID_COMMAND: "INVALID_COMMAND",
	COMMAND_EXECUTION_ERROR: "COMMAND_EXECUTION_ERROR",
	STREAM_START_ERROR: "STREAM_START_ERROR",
	PROCESS_NOT_FOUND: "PROCESS_NOT_FOUND",
	PROCESS_PERMISSION_DENIED: "PROCESS_PERMISSION_DENIED",
	PROCESS_ERROR: "PROCESS_ERROR",
	SESSION_ALREADY_EXISTS: "SESSION_ALREADY_EXISTS",
	PORT_ALREADY_EXPOSED: "PORT_ALREADY_EXPOSED",
	PORT_IN_USE: "PORT_IN_USE",
	PORT_NOT_EXPOSED: "PORT_NOT_EXPOSED",
	INVALID_PORT_NUMBER: "INVALID_PORT_NUMBER",
	INVALID_PORT: "INVALID_PORT",
	SERVICE_NOT_RESPONDING: "SERVICE_NOT_RESPONDING",
	PORT_OPERATION_ERROR: "PORT_OPERATION_ERROR",
	CUSTOM_DOMAIN_REQUIRED: "CUSTOM_DOMAIN_REQUIRED",
	GIT_REPOSITORY_NOT_FOUND: "GIT_REPOSITORY_NOT_FOUND",
	GIT_BRANCH_NOT_FOUND: "GIT_BRANCH_NOT_FOUND",
	GIT_AUTH_FAILED: "GIT_AUTH_FAILED",
	GIT_NETWORK_ERROR: "GIT_NETWORK_ERROR",
	INVALID_GIT_URL: "INVALID_GIT_URL",
	GIT_CLONE_FAILED: "GIT_CLONE_FAILED",
	GIT_CHECKOUT_FAILED: "GIT_CHECKOUT_FAILED",
	GIT_OPERATION_FAILED: "GIT_OPERATION_FAILED",
	BUCKET_MOUNT_ERROR: "BUCKET_MOUNT_ERROR",
	S3FS_MOUNT_ERROR: "S3FS_MOUNT_ERROR",
	MISSING_CREDENTIALS: "MISSING_CREDENTIALS",
	INVALID_MOUNT_CONFIG: "INVALID_MOUNT_CONFIG",
	INTERPRETER_NOT_READY: "INTERPRETER_NOT_READY",
	CONTEXT_NOT_FOUND: "CONTEXT_NOT_FOUND",
	CODE_EXECUTION_ERROR: "CODE_EXECUTION_ERROR",
	PYTHON_NOT_AVAILABLE: "PYTHON_NOT_AVAILABLE",
	JAVASCRIPT_NOT_AVAILABLE: "JAVASCRIPT_NOT_AVAILABLE",
	OPENCODE_STARTUP_FAILED: "OPENCODE_STARTUP_FAILED",
	PROCESS_READY_TIMEOUT: "PROCESS_READY_TIMEOUT",
	PROCESS_EXITED_BEFORE_READY: "PROCESS_EXITED_BEFORE_READY",
	VALIDATION_FAILED: "VALIDATION_FAILED",
	INVALID_JSON_RESPONSE: "INVALID_JSON_RESPONSE",
	UNKNOWN_ERROR: "UNKNOWN_ERROR",
	INTERNAL_ERROR: "INTERNAL_ERROR"
};

//#endregion
//#region ../shared/dist/errors/status-map.js
/**
* Maps error codes to HTTP status codes
* Centralized mapping ensures consistency across SDK
*/
const ERROR_STATUS_MAP = {
	[ErrorCode.FILE_NOT_FOUND]: 404,
	[ErrorCode.COMMAND_NOT_FOUND]: 404,
	[ErrorCode.PROCESS_NOT_FOUND]: 404,
	[ErrorCode.PORT_NOT_EXPOSED]: 404,
	[ErrorCode.GIT_REPOSITORY_NOT_FOUND]: 404,
	[ErrorCode.GIT_BRANCH_NOT_FOUND]: 404,
	[ErrorCode.CONTEXT_NOT_FOUND]: 404,
	[ErrorCode.IS_DIRECTORY]: 400,
	[ErrorCode.NOT_DIRECTORY]: 400,
	[ErrorCode.INVALID_COMMAND]: 400,
	[ErrorCode.INVALID_PORT_NUMBER]: 400,
	[ErrorCode.INVALID_PORT]: 400,
	[ErrorCode.INVALID_GIT_URL]: 400,
	[ErrorCode.CUSTOM_DOMAIN_REQUIRED]: 400,
	[ErrorCode.INVALID_JSON_RESPONSE]: 400,
	[ErrorCode.NAME_TOO_LONG]: 400,
	[ErrorCode.VALIDATION_FAILED]: 400,
	[ErrorCode.MISSING_CREDENTIALS]: 400,
	[ErrorCode.INVALID_MOUNT_CONFIG]: 400,
	[ErrorCode.GIT_AUTH_FAILED]: 401,
	[ErrorCode.PERMISSION_DENIED]: 403,
	[ErrorCode.COMMAND_PERMISSION_DENIED]: 403,
	[ErrorCode.PROCESS_PERMISSION_DENIED]: 403,
	[ErrorCode.READ_ONLY]: 403,
	[ErrorCode.FILE_EXISTS]: 409,
	[ErrorCode.PORT_ALREADY_EXPOSED]: 409,
	[ErrorCode.PORT_IN_USE]: 409,
	[ErrorCode.RESOURCE_BUSY]: 409,
	[ErrorCode.SESSION_ALREADY_EXISTS]: 409,
	[ErrorCode.SERVICE_NOT_RESPONDING]: 502,
	[ErrorCode.GIT_NETWORK_ERROR]: 502,
	[ErrorCode.PYTHON_NOT_AVAILABLE]: 501,
	[ErrorCode.JAVASCRIPT_NOT_AVAILABLE]: 501,
	[ErrorCode.INTERPRETER_NOT_READY]: 503,
	[ErrorCode.OPENCODE_STARTUP_FAILED]: 503,
	[ErrorCode.PROCESS_READY_TIMEOUT]: 408,
	[ErrorCode.PROCESS_EXITED_BEFORE_READY]: 500,
	[ErrorCode.NO_SPACE]: 500,
	[ErrorCode.TOO_MANY_FILES]: 500,
	[ErrorCode.TOO_MANY_LINKS]: 500,
	[ErrorCode.FILESYSTEM_ERROR]: 500,
	[ErrorCode.COMMAND_EXECUTION_ERROR]: 500,
	[ErrorCode.STREAM_START_ERROR]: 500,
	[ErrorCode.PROCESS_ERROR]: 500,
	[ErrorCode.PORT_OPERATION_ERROR]: 500,
	[ErrorCode.GIT_CLONE_FAILED]: 500,
	[ErrorCode.GIT_CHECKOUT_FAILED]: 500,
	[ErrorCode.GIT_OPERATION_FAILED]: 500,
	[ErrorCode.CODE_EXECUTION_ERROR]: 500,
	[ErrorCode.BUCKET_MOUNT_ERROR]: 500,
	[ErrorCode.S3FS_MOUNT_ERROR]: 500,
	[ErrorCode.UNKNOWN_ERROR]: 500,
	[ErrorCode.INTERNAL_ERROR]: 500
};

//#endregion
export { ErrorCode as t };
